File: algorithm/sorting/python/quick_sort.py
def swap(list, index1, index2):
    #print("swap " + str(list[index1]) + " and " + str(list[index2]))
    temp = list[index1]
    list[index1] = list[index2]
    list[index2] = temp
    #print(list)

def quick_sort(list, begin, end):
    if begin >= end:
        return
    left = begin
    right = end
    first = list[begin]
    while True:
        while list[right] >= first and left < right:
            right -= 1
        while list[left] <= first and left < right:
            left += 1
        if left == right:
            break
        swap(list, left, right)
    swap(list, begin, left)
    quick_sort(list, begin, left - 1)
    quick_sort(list, left + 1, end)

if __name__ == "__main__":
    list = [3, 2, 7, 4, 5, 9, 1, 8, 0, 6,]
    quick_sort(list, 0, len(list) - 1)
    print(list)

File: algorithm/sorting/cpp/quick_sort.cpp
#include <iostream>
using namespace std;

void swap(int* a, int* b){
    int temp = *a;
    *a = *b;
    *b = temp;
    cout << "swap " << *a << " and " << *b << endl;
}

void quickSort(int* start, int* end){
    if(start >= end)
        return ;
    int pivot = *start;
    int* left = start;
    int* right = end;
    while(left < right){
        while(*right > pivot && left < right)
            right -= 1;
        while(*left <= pivot && left < right)
            left += 1;
        if(left == right)
            break;
        swap(left, right);
    }
    swap(left, start);
    quickSort(start, left - 1);
    quickSort(left + 1, end);
}

void print(int* arr, int len){
    while(len--){
        cout << *arr << " ";
        arr += 1;
    }
    cout << endl;
}

int main(int argc, char const *argv[])
{
    int numList[10] = {6, 1, 2, 7, 9, 3, 4, 5, 10, 8, };
    quickSort(numList, numList + 9);
    print(numList, 10);
    return 0;
}